### ISSUE TITLE
Add the ability to schedule publishing of content in the future.

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#857]
A feature WordPress users really enjoy is the ability to schedule a post to go live at 6am and not have to wake up at 6am to push the publish button.  It would be nice if we could do something similar in core.  

I think it would be reasonable to use the post-date for this, if the date is in the future, schedule the post to go live at that date.  If not, then publish right away.

However, if we want editor-types to be able to schedule a post, they will also need access to the post-date field, something that is usually only granted with "administer nodes" permission. See related: https://github.com/backdrop/backdrop-issues/issues/815#issuecomment-89483227

---

PR by @docwilmot: https://github.com/backdrop/backdrop/pull/2139

**Comment by pablo-romero** [comment]
+1, I was about to suggest it too.

There's also a problem in the current scenario:

If you set a future date (even by mistake) in the **Authored on** field when creating or editing content, that content gets published straight away (if **Published** checkbox is checked, which is by default).
Not sure if anyone has ever used it that way on purpose, but the result never made much sense to me: having a visible submitted-on-the-future content, e.g. submitted on 05/01/2015 but last updated on 04/16/2015. It also leads to confussion, specially for those who usually work with WP.

Maybe we should also consider the "Published" checkbox status, to allow users to save unpublished contents ("drafts") with future dates even when they don't need automatic publication.

Some conditional like:

* If "Authored on" is in the past, keep the content published/unpublished (depending on the "Published" checkbox status) with the custom datetime.
* Else if "Authored on" is blank, keep the content published/unpublished (depending on the "Published" checkbox status) and set the current datetime.
* Else if "Authored on" is in the future:
  * If "Published" is unchecked: save the content (unpublished with a custom future "Authored on" datetime), and do not publish it until "Publish" checkbox is checked.
  * If "Published" is checked: keep the content unpublished (hidden), and publish it automatically on the custom future "Authored on" datetime (using the *publish_content* action with cron).

BTW, the [Scheduler Module](https://drupal.org/project/scheduler) for Drupal does it all: It has its own permissions and also lets you automatically unpublish content on specific dates.
It could be a good reference.

**Comment by klonos** [comment]
...adding this to #1388:

![wp-content_scheduled_publishing](https://cloud.githubusercontent.com/assets/2423362/12071863/d2d350a0-b116-11e5-8426-28e659d41368.png)
![wp-content_scheduled_publishing-options](https://cloud.githubusercontent.com/assets/2423362/12071864/d8ea4a98-b116-11e5-9c2c-ab4fabc03496.png)

**Comment by klonos** [comment]
...it's not "scheduling" in quotes - that's what it actually is.
...and it's for all content - not just posts.

**Comment by jenlampton** [comment]
Yes. We already have published date. If someone sets a post to draft (or whatever we call it) and sets a published date in the future, then it should publish at that date! :+1:

**Comment by serundeputy** [comment]
@jenlampton  have a look at this.

I've taken a stab at this.
Functionally works.
In general feedback is welcome.
In specific the help text on the form fields could use some feedback.

![screen shot 2016-02-20 at 11 07 17 am](https://cloud.githubusercontent.com/assets/6763671/13197675/cd1c4432-d7c2-11e5-8811-2e9cc1081b13.png)

`PR` https://github.com/backdrop/backdrop/pull/1262

Happy reviewing :smile:

**Comment by quicksketch** [comment]
Thanks @serundeputy! It's a good start. It's nice to see that the actual code for scheduling is so straight-forward. Just `hook_cron()` and publish the nodes. :smile:

The UI component is definitely going to be the trickiest part here. @wesruv or @dariusgarza might be able to provide some UX feedback here, but I think we need to design the UI here first and then implement it. The reuse of common patterns (collapsed fieldsets) is familiar but not the best fit here.

A couple of thoughts right off:

- Let's skip any mention of cron jobs here. Though I think it would be appropriate for us to shorten the default cron frequency from 3 hours to something like every 10 minutes. That would ensure scheduling works predictably.
- We need more granularity than date alone, it's not uncommon to schedule something for 8am or some other time than midnight.

**Comment by serundeputy** [comment]
Thanks!

I'll wait for UX feedback/direction before proceeding.

**Comment by serundeputy** [comment]
updated `PR` to address some of the issues brought up by @quicksketch 

> Let's skip any mention of cron jobs here. Though I think it would be appropriate for us to shorten the default cron frequency from 3 hours to something like every 10 minutes. That would ensure scheduling works predictably.
We need more granularity than date alone, it's not uncommon to schedule something for 8am or some other time than midnight.

Added cron config for running every 10min, 30min or 1hr, 3hr etc.
<img width="1127" alt="screen shot 2016-02-20 at 8 12 52 pm" src="https://cloud.githubusercontent.com/assets/6763671/13200262/67ff8ba8-d80e-11e5-84cb-59c84eb5d986.png">

Added scheduling granularity to correspond to the 10min interval:
<img width="1271" alt="screen shot 2016-02-20 at 8 12 18 pm" src="https://cloud.githubusercontent.com/assets/6763671/13200266/82243d62-d80e-11e5-9b3f-67d0f0d3e8bb.png">

Need UX review followed by implementation of UX recommendations. @wesruv , @dariusgarza, @jenlampton other UX and/or any other UX opinions please review if you get a chance!

thanks,
~Geoff

**Comment by docwilmot** [comment]
I'm not one of those designers but I'll weigh in: 
I'd go with radios for the options to publish (now vs later) and reveal the date picker using states.

**Comment by wesruv** [comment]
I would love to take a crack at this! 

If I forget to, someone badger me.

**Comment by klonos** [comment]
Thanx @serundeputy for taking a crack at this!

> I'm not one of those designers but I'll weigh in:
I'd go with radios for the options to publish (now vs later) and reveal the date picker using states.

:+1: you beat me to it @docwilmot! It was the first thing that caught my eye as weird UX: what happens if someone checks "Publish now" option? ...does the fieldset go away (#states)? ...if so, why not have that as the default option? ...what is that "Yes" checkbox within the fieldset? Anyways, as the Doc said, a radio set is in order here.

> - Let's skip any mention of cron jobs here....

Also was about to comment on that. Same situation as mentioning "node" in the front end. No-no.

> ...Though I think it would be appropriate for us to shorten the default cron frequency from 3 hours to something like every 10 minutes. That would ensure scheduling works predictably.

It feels that this deserves its own issue. I know that we have background cron runs now (#1214), but what happens if all the cron jobs take more than the set interval (more chances to hit that use case if we shorten the time to 10mins)? Did we account for that scenario in https://github.com/backdrop/backdrop/pull/1193 ?

**Comment by klonos** [comment]
...for the record, I'm all in in shortening the time of cron runs ...especially now that the website is served to users no matter what state the cron is in. It's just that we need to think what happens if cron runs almost constantly in the background (in terms of say CPU and hard disk activity etc.)?

**Comment by gifad** [comment]
> now that the website is served to users no matter what 

Beware that this is a marketing claim, usefull only for the external visitor first visit.
If you are a regular user (contributor), you will notice that your 2/4 first clicks to go to the page you want to edit are painfully slow - again for a real site, with dozens of contributed modules installed.

Also, every morning, I get the "Attempting to re-run cron while it is already running." message 2..4 times, because any site access while cron is running launch a new cron run...

**Comment by serundeputy** [comment]
Thanks everyone.  You all have good points.

I just want to clear up the current state of affairs concerning use of states and the use cases we need to cover.

I made a vid so you can see things in action: https://youtu.be/5Xu2Bga0pdI

**Correction**: I should have said current time exceeds pulished_on.

The crux of the matter is that we need to account for three states:
  1.  User wants to publish immediately
  2.  User wants to save a draft and publish manually at their discretion.
  3.  User wants to schedule future publication

@wesruv :smile: I'm glad you will take on the UX, just make sure we account for the 3 states.  So exciting !

thanks everyone.

**Comment by Graham-72** [comment]
:+1:  This will be really useful. And thanks for the clear youtube demo.

**Comment by wesruv** [comment]
Here's my quick take, a few notes/caveats:
* I need to think about Promoted/Sticky at top of lists more... they're odd ducks, they don't even feel like publish options IMO. I'll put some more thought into they're appearance in the form, just slapped them on for now.
* I think we should use a consistent time field widget. Right now Authoring Info is using the widget that I've mocked up, I don't see any good reason why they shouldn't be the same. I vote we change/improve both, or use the same one here. I do like the explicitness of this field widget, even if it is a little unsightly.
* On Android's Alarm app, when you set the alarm the confirmation says "Alarm set for 8 hours 37 minutes from now" (for instance). This has saved my butt from confusing AM/PM multiple times. Thinking a similar notification for this would be helpful. Tomorrow, Next $DAY_NAME, $N Months, etc. Don't know how reasonable that is from an implementation perspective.

## If default for the content type is to publish now
![image](https://cloud.githubusercontent.com/assets/5607236/13204024/1433ed90-d893-11e5-91d8-8d6f457a5e5b.png)

## Open dropdown
![image](https://cloud.githubusercontent.com/assets/5607236/13204032/1d73bd68-d893-11e5-9a0f-59db60971802.png)

## Chose Schedule Publish Time
![image](https://cloud.githubusercontent.com/assets/5607236/13204040/38a401a6-d893-11e5-8d6d-72a55e9e7689.png)

## Clicked on Schedule Time Field
![image](https://cloud.githubusercontent.com/assets/5607236/13204051/95a60e44-d893-11e5-99f6-70f36938480b.png)

## Left field and notification telling user how far off publish date is appears
![image](https://cloud.githubusercontent.com/assets/5607236/13204056/ba539932-d893-11e5-9b4c-bd955bf40fd9.png)

**Comment by gifad** [comment]
This is nice, but does anybody have an idea on how to extend this to "schedule publishing of new revision of content in the future" ?
Yes, I know, it's much more complicated… (see [Support revisions in different states](https://www.drupal.org/node/218755))

**Comment by klonos** [comment]
Thoughts...

@wesruv :+1: on being consistent with the date selection UI and also big :+1: on the scheduling notification! **But**...

Using a drop-down requires two clicks in order for a user to select an option. Using radio buttons on the other hand requires only a single click.

I understand that the publishing options are currently in the first, selected by default vtab, but if for some reason in the future we introduce another vtab and make that the default, then we hide these options not only inside the drop-down but also bury them in side a vtab on top of it. So add another click there.

Also, drop-down menus are perfect for simplifying the UI when there are lots of options. With publishing though, we only intent to have 3 options currently. Putting them in a radio set instead would allow users to discover the available options without any additional clicks. Better UX if you ask me.

@wesruv:

> I need to think about Promoted/Sticky at top of lists more... they're odd ducks, they don't even feel like publish options IMO.

I don't agree with you on this one. I think they *are* publishing options ...**BUT**, I hardly ever use them. So if you meant to say that, then I'm with you. You see most of my sites have custom front pages, so "Promote" for example is moot. I don't use any blog/forum-style pages either where the "Sticky" option would be useful. So, I end up removing these options from the node edit forms and basically, the "Published" checkbox is the only thing in the vtab:

![backdrop-node_edit_form-publishing_options_vtab-only_published_checkbox](https://cloud.githubusercontent.com/assets/2423362/13206104/d85de3a2-d94b-11e5-9228-ba06c3efa9d5.png)

I'm sure you'll agree that this is a total waste of UI space. And I don't think that my use case (disabling the "Promote" and "Sticky" options) is a rare one at all. If anything else, people might simply just be leaving these options there because they are the default and they can't be bothered to remove them.

Anyways, my point is that the way I see it, "Publish" (now) and "Save" (draft) are both one-click actions and will be the most commonly used. These can simply be form actions instead of being buried away in a vertical tab (and even more inside a drop-down). Now the "Schedule" option needs more input from the user (date selection), so that's a different story.

In #1475 I mentioned this:

> Why do we keep the "Published" option in a vtab? Couldn't we simply move this checkbox in the form of a separate "Publish" button (and change the "Save" button to something like "Save as draft")?

@gifad since you mentioned revisions in different states, if you haven't read #1475 already, I suggest you take a look. :wink:

**Comment by klonos** [comment]
I think that the scheduling feature that we intent to implement here will be a very useful one and I intent to use it extensively in sites where it makes sense. It also helps us come closer to feature parity with WP, so :+1:  ...**but** I don't think that it will be a very commonly used action for the average website. So, I propose we go with something similar to what they did in D8:

![d8-save_node_options](https://cloud.githubusercontent.com/assets/2423362/12071116/71ef5bf8-b0ee-11e5-97e7-cb67b3611f9c.png)

...but we have the options in the drop button be "Publish now", "Schedule publishing..." and "Unpublish" (or "Save and unpublish" that although a bit verbose, is more descriptive).

If the user selects the "Schedule publishing..." option, they get a dialog where they select date and time and the "This [node-type-name] is scheduled for publishing on [date-day-or-tomorrow] at [time]" notification is shown in the node view page (once saved).

**Comment by quicksketch** [comment]
> ...but we have the options in the drop button be "Publish now", "Schedule publishing..." and "Unpublish" (or "Save and unpublish" that although a bit verbose, is more descriptive).

I have a strong aversion to using a drop button as the primary action button within a form. The D8 save button is the only UI I have ever seen that used a dropbutton for the main action. Using a drop button for "Schedule publishing..." would have new challenges. You'd end up clicking the primary button once, but then having something else ask for the scheduling details (a dialog?) and then have to click again to actually execute saving the form.

**Comment by klonos** [comment]
...thinking ahead here for when we implement a workflow system: we can have a "Publish, schedule and unpublish content" permission that toggles the availability of this drop-down for users with that permission specifically. The rest of the users will be limited to the options of another drop button with options like "Save draft" and "Save and request review" or something like that (#1475).

**Comment by wesruv** [comment]
I totally agree, the dropbutton is not a common design pattern, and the cost of a misclick if you meant to save a draft is too high for my taste.

Also having scheduling come off of that would get reeeaallyy awkward (at least the way I'm picturing it).

**Comment by wesruv** [comment]
@klonos good point that the design should cover the different situations for perms/settings on the content type.
* Only has publish
* Has Draft or Publish
* Has draft, publish and schedule publish

**Comment by klonos** [comment]
@quicksketch I hear you, but as I said, I don't think that the scheduling option will be the norm for average websites. Hence, my suggestion to be hidden as an option of a drop button. Nevertheless, if drop buttons is a no-no for primary submit action in forms, then how about a separate "Schedule publishing..." button?

**Comment by klonos** [comment]
...my main point is to lose the "Publishing options" vtab and this argument specifically:

>Why do we keep the "Published" option in a vtab? Couldn't we simply move this checkbox in the form of a separate "Publish" button (and change the "Save" button to something like "Save as draft")?

**Comment by gifad** [comment]
@klonos : D8 "Save as unpublished" is not the answer to my question : It unpublishes the node and all its revisions !
I'd like to save my edits, not publish the last revision, but keep the last published revision 'active'.
But I'm getting off topic :confused:

**Comment by wesruv** [comment]
I'm going to stick to the scope of this ticket, keep Published where it is in the vertical tabs, and not worry about Revision scheduling/publishing for now. Although I am interested in both of those topics.

## Default state
![image](https://cloud.githubusercontent.com/assets/5607236/13207362/a5566580-d8db-11e5-8201-4429cad02217.png)

## Schedule Publish
![image](https://cloud.githubusercontent.com/assets/5607236/13207382/e2059d8e-d8db-11e5-9a09-0bdb5a81385f.png)

## When a page is published
![image](https://cloud.githubusercontent.com/assets/5607236/13207406/14cc8c6e-d8dc-11e5-874c-de4acdf4ce95.png)

If the user unpublishes the content it'd go back to the 'default state' mockup.

**Comment by serundeputy** [comment]
new `PR` meets most of @wesruv 's suggestions.  Need some work on the `js` when you click in the date field it adds some to the vtab weirdly :hammer: 

Taking away the -[ ] Published check box changes the complexity of the additional code.  I'll be curious to get a review @quicksketch ; I've put in some helper functions to remember when in `edit` mode what the published date is and if the node is in draft or scheduled mode.

that is all for tonight.

**Comment by serundeputy** [comment]
hehe; @wesruv you beat my post, but yeah; I should be able to change to radios without too much trouble.

I definitely agree with @wesruv assessment of the scope of this ticket.

I am also interested in the other topics, workflows and revisions while keeping a published version, but that is a new ticket I think.

~Geoff

**Comment by klonos** [comment]
Fair enough about not getting of topic here, but the "Unpublish" button in the vtab is as weird a UI pattern as the dropbutton. Perhaps move between "Save" and "Delete" would make more sense. The "This content is published" status text can remain in the vtab.

@gifad I pointed out #1475 to you because the d.org issue about revisions in different states was superseded by the follow up issue about arbitrary workflow states. I mention that there, so yes, although not exactly of topic because the issues are related, we are indeed sidetracking this issue. Lets keep the discussion about advanced workflows in #1475.

**Comment by docwilmot** [comment]
I like @wesruv mockups. Havent tried the PR.

**Comment by klonos** [comment]
...and since we are implementing publishing something in the future (announcements of products getting end-of-life status or "Happy new year!" nodes for example), we should consider implementing unpublishing content too. For instance product offers that start at a specific point in time and also end in a specific date.

**Comment by serundeputy** [comment]
Updated `PR` with radio buttons fixed js.
`PR` https://github.com/backdrop/backdrop/pull/1262
`patch` https://github.com/backdrop/backdrop/pull/1262.patch

Still need to fix tests as the original checkbox for published was removed.

@klonos that seems like a fine idea.  should we do it here, a separate issue.  What do others think.  Anybody second?  We would need to get the UX for `schedule unpublish` too.

**Comment by docwilmot** [comment]
> We would need to get the UX for schedule unpublish too. 

Separate issue, I think. We'd need to be sure about the process and implications.

**Comment by klonos** [comment]
> ...that seems like a fine idea. should we do it here, a separate issue.

Well, not my idea exactly. When I make my suggestions in this issue here, what I have in mind is feature parity with https://www.drupal.org/project/scheduler:

> Scheduler gives content editors the ability to schedule nodes to be published **and unpublished** at specified dates and times in the future.

(emphasis mine)

![drupal-contrib-scheduler-publish_and_unpublish_options](https://cloud.githubusercontent.com/assets/2423362/13237117/59c30270-da1d-11e5-8446-ce28019b8176.png)

As I said, I don't think that the average Joe will be using this entire feature of scheduled publishing/unpublishing in their website (that specific module has only about 68k reported installations at the time of writing). Nevertheless, it is a great feature, it brings us in parity with WordPress and it will be a good addition to our marketing arsenal (for websites that do article publishing for example which is a good market share and also e-commerce sites where they need to schedule publishing and unpublishing of product specials).

I don't mind filing separate issues and making this a meta-issue, but in order to better design the UI and brainstorm on the UX and workflow, I think that it'd be best to tackle this here and take all aspects of it under consideration. Otherwise, we'll be risking to make UI design decisions that we might need to change/adapt as we add features in. Why do double the trouble instead of getting the design right for every sub-feature that we intent to implement in a single go?

Having said all that, I do not have the time to do so right now, but I'll pt a list together of a couple of things that we need to keep in mind and some shortcomings of the design of the d.org contrib module that either do not apply to us or I have better ideas for...

**Comment by serundeputy** [comment]
just realized that these need to be addressed as well;

* /admin/structure/types/add
* /admin/structure/types/manage/{type}/configure

By removing the `Published` check box the `Publishing settings` vtab will not work in its current form.

**Comment by klonos** [comment]
@serundeputy that's one of the thing I had in mind when I said:

> ...I'll put a list together of a couple of things that we need to keep in mind and some shortcomings of the design of the d.org contrib module that either do not apply to us or I have better ideas for...

:wink:

**Comment by klonos** [comment]
...So here goes:

1. [Scheduler](https://www.drupal.org/project/scheduler) has implemented a separate "Scheduler" vtab in the **content type** settings to the "Publishing options" vtab that holds the "Published" checkbox:

![drupal-contrib-scheduler-content_type_form](https://cloud.githubusercontent.com/assets/2423362/13240530/566150fe-da37-11e5-968d-e6d22a7e451f.png)

This way, in the **node add/edit form** the defaults are set so that the node is published *and* there is also a separate vtab in that form as well for the publish/unpublish options:

![drupal-contrib-scheduler-content_add-edit_form](https://cloud.githubusercontent.com/assets/2423362/13240558/9d06f66c-da37-11e5-8533-52cd9ca7da90.png)

So if the user forgets to uncheck that option, this effectively creates a WTF situation where while the nodes of that content type are scheduled to be published at a later time, they are published immediately upon node save.

This issue does not apply to backdrop since we will be placing the scheduling options in the same tab on the content type form **and** the content add/edit form. The user will have a radio set available to them and they can only select **one of** the options: either "Publish" (now) or "Schedule".

The same applies for default settings for the content type form and we'll need to make sure that these are also a radio set so that the default options are "Publish immediately" or "Save draft" or "Schedule for publishing" and (IMO) "Publish immediately" is the default.

**Comment by klonos** [comment]
...having the default option of the content type set to "Schedule for publishing" will effectively make the scheduling thing a requirement because that will be the default when a user clicks the form submit button. If we have the publishing date/time fields be empty by default, that will trigger a validation error (while still giving the user the chance to select "Publish now"). This could either be considered a feature or a bug. In the later case, we'll need to think this option a bit more...

Perhaps show only the two options ("Publish immediately" / "Save *unpublished* draft") and have the third one for scheduling only available when the admin selects a "Enable scheduled publishing for this content type" checkbox. Then we also show a "Require scheduled publishing" checkbox? ...for reference, here's what the Scheduler respective options look like...

Before enabling the options for publishing/unpublishing:

![drupal-contrib-scheduler-content_type_form-publishing-unpublishing_options_unchecked](https://cloud.githubusercontent.com/assets/2423362/13241085/31bcb18a-da3c-11e5-9408-e95f4ee4896d.png)

...and after these options are enabled:

![drupal-contrib-scheduler-content_type_form-publishing-unpublishing_options_checked_and_advanced_options_expanded](https://cloud.githubusercontent.com/assets/2423362/13241088/38e493ba-da3c-11e5-8195-3b88342219d2.png)

- the "Change content creation time to match the scheduled publish time" option that I think would be useful to offer as well
- the "Create a new revision on publishing" option, I don't think we need since we can merge it with the one we have in the publishing options tab
- the advanced section for handling dates in the past, I'm not sure what their use is since I'd go with "Show error message - Do not allow" and not give any option to the user. I could be wrong though and there might be a use case for it.
- the "Node edit page layout" again I say we ditch and go with the "Publishing options" vtab.

**Comment by serundeputy** [comment]
I've updated the `PR` https://github.com/backdrop/backdrop/pull/1262

It includes the ability to set corresponding publishing state defaults to the node_type form `admin/structure/types/add` and the corresponding configure forms:

<img width="878" alt="screen shot 2016-02-28 at 7 35 18 pm" src="https://cloud.githubusercontent.com/assets/6763671/13383124/79a4eaca-de52-11e5-98bf-9bcf2682d8e8.png">

I've added tests in `node.test` The cases for `Draft` and `Publish Now` are passing fine, but the case for `Schedule Publish Time` is producing an error:

<img width="1274" alt="screen shot 2016-02-28 at 7 24 26 pm" src="https://cloud.githubusercontent.com/assets/6763671/13383135/b61c45a2-de52-11e5-8725-388e7b522ac8.png">

Same error in text:

```
strtotime() expects parameter 1 to be string, array given	Warning	node.pages.inc	512	
```

It seems to me that the value `$form_state['values']['future-publish']` passed to `strtotime()` is a string.  I did some digging on this @quicksketch do you think this could be related: https://www.drupal.org/node/2051921 ?

I'd like to get that test sorted out.

thanks,
~Geoff

**Comment by BWPanda** [comment]
Some random thoughts after skimming over this issue (forgive me if I've missed where these have already been discussed):

- I propose naming the radio options: Publish now, Publish later and Save draft (in that order - I think 'Publish now' should be the first and default option)
- Help text/description should be provided explaining what each option means/does
- Has any consideration been given to how nodes change between published/scheduled states? E.g. I understand that scheduled nodes are stored as 'unpublished' in the database? What happens if they're 'Published' before that date (either from the DB directly, or via a contrib module)? What happens if a published node is marked as 'scheduled'?

**Comment by jenlampton** [comment]
> I propose naming the radio options: Publish now, Publish later and Save draft (in that order - I think 'Publish now' should be the first and default option)

I agree that `Publish now` should be the first and default option, but as a user I don't understand the difference between `Save draft` and `Publish later`. I'll certainly want to publish this post later if I am saving is as a draft now, those two options are likely to get confused.

I think we need to make it clear that one of these options involves scheduling the post to be published in the future, so using the word `Schedule` is key. I looked at the examples above, and I don't really like either `Schedule publish` or 'Schedule publish time`. I agree that we don't need the word `time` if we're using the word `schedule` but I also think we don't really need the word `Publish`.  What if we used...

* Publish now
* Schedule for later
* Save draft

**Comment by jenlampton** [comment]
> What happens if they're 'Published' before that date (either from the DB directly, or via a contrib module)?  What happens if a published node is marked as 'scheduled'?

From looking at the code, nothing. Only posts that re are both unpublished and scheduled will be published at the specified time. (and I believe this is the expected behavior)

**Comment by BWPanda** [comment]
> I like this because most of the time developers looking for a published date will grab $node->published instead of $node->created -- and they would be correct :)

Does that mean that when publishing a node directly (without scheduling), that $node->published will be the same as $node->created? I don't think we want a scenario where $node->published is empty (that might be confusing)...

**Comment by docwilmot** [comment]
I wanted to get this workable before the weekend is over, but I suspect we'll need to have conversations before we settle on all the PR comments so I doubt we'll all agree by today, so I hope @serundeputy can carry it over the finish line if I dont.
I agree with the majority of the comments but I really dont like calling the new DB column `published`, thats way too confusing next to `status`. How about just `published_date`?

**Comment by jenlampton** [comment]
> I really dont like calling the new DB column published, thats way too confusing next to status. 

I had the same concern initially, but when I thought it through more it stopped bothering me for the following reasons: 

1) status is boolean, but all the past tense verb fields would be dates (created, updated, published), this is nice consistency.
2) we're not changing `status` at all, so there will be no change for people already familiar with backdrop (or Drupal)
3) devs who are new to backdrop and querying the database would certainly do a DESC on the table  to get the column name anyway, and in doing so they would see which field was which.
4) I can think of a handful of reasons both core and contrib should be using the `published` date instead of the `created` date (for example, in the "submitted" information on nodes) and by adding a `published` column we'd be making it very clear to folks that this is that value they've always wanted.  Calling it something like "scheduled" (see below) wouldn't have the same "aha!" effect.

Another idea I had was to call it `scheduled` instead.

The down side do that is that it wouldn't be clear that the `scheduled` column held the published date (for nodes that had already been published) as well as the scheduled date (for nodes yet-to-be-published). Published is just a more accurate name for what's in the column.

> How about just published_date?

I could get behind this if we also changed the column names from `updated` to `updated_date` and `created` to `created_date`, but I'm not sure that's something we'd want to do in the 1.x cycle. 

For reference, in WordPress the columns are named `POST_DATE` and `UPDATED_POST_DATE` (and they are also datetime fields). Interesting that WP not have a concept of a created date.

**Comment by jenlampton** [comment]
> the published/posted date

To me the word "posted" means the same thing as "created" and we shouldn't confuse that with the date a post was published (or scheduled to be published).

**Comment by BWPanda** [comment]
> What I'm expecting to see out of this is that "created" starts being the actual creation date (and would likely become unchangeable, like "updated" dates are). And the new "published" date would be the date used when displaying a piece of content, scheduling the piece of content, and shown in the node form. Am I being too ambitious here?

Not at all, I love that idea! I just wrote a blog post the other day that took me a few hours (saving as draft every so often), and when I published it, I manually changed the date to match the current time (I'm pedantic like that).

So it makes sense that 'created' is when `node/add` is first submitted (in whatever state), and 'published' is just that: when the node first enters published state (maybe as soon as it's created, or maybe later when scheduled or manually published from draft).

**Comment by jenlampton** [comment]
My vote is also for `published` for the database column name, for all the [reasons listed above](https://github.com/backdrop/backdrop-issues/issues/857#issuecomment-383427757), but mostly for #4: When developers look at the columns available to them in the database searching for the published date (and this is a very common use case), they won't recognize it instantly if the column name is `scheduled` (or `created`).

For the node itself, I think we can improve DX by having both a `$node->published` that will be 0 or FALSE before the node is published, and a `$node->scheduled` that will be the published date before the node is published. This way if a developer does accidentally grab `$node->published` instead of `$node->status` they'll get the same answer (FALSE before the node is published and TRUE after) as if they had grabbed `$node->status`.

> Am I being too ambitious here?

I don't think so. I am also really starting to love the idea of eliminating the `created` date from the UI.  Having a `published` date is what people really want, and using that for scheduling too will clean up the UI.

**Comment by quicksketch** [comment]
Thanks @BWPanda and @jenlampton for the feedback. There is also yet another option, one I am a little hesitant to suggest but would be the least work of all:

- If we *don't* change how `created` is used, and we continue showing it as the submitted by line, using it in administrative listings, and expose it on the node form, then `created` in practice doesn't mean "created" at all, it means "published_date".
- That means that if you were to treat "created" as "published" universally, it means that we could schedule the publish date in the future using this same field, akin to what WordPress does. They only have "post_date", which if is in the future, is used for scheduling.
- Then we don't need a new database column at all. We don't need to update any existing views. We don't need to change the submitted-by line. We don't need to change any APIs.
- But the downside of it all is that "created" doesn't mean "created" at all. It means "published", and we just have a very poorly named database column and node property (essentially continuing status-quo from Drupal 7).

We'd have to wait until Backdrop 2.x, at which time we could theoretically rename "created" to "published_date" (and make it a DateTime column in MySQL instead of an integer timestamp). Though that level of breakage also makes me nervous even suggesting.

All of this is a hypothetical. I'm not much a fan of it, but it would be even less complex than what has been done here. However, the fact is we have the current code working and it's a great implementation. We'll just need more follow-ups to start using the new "published" date everywhere.

What I like least of all though, is using this new column *only* for scheduling, and _continuing_ to abuse the `created` column by updating its timestamp when content is published. Then we have not only an inaccurate column, but a new one that is unnecessary. If we're going to abuse "created" and treat it as the published date, we should commit entirely to that approach and also use it for scheduling into the future. Or if we want to introduce a published date, we should use it universally as well (in follow-up issues after this one), replacing nearly every place "created" is currently used.

**Comment by BWPanda** [comment]
Um, I'm starting to get confused... I don't know what explanations are expanding upon existing suggestions and what are adding new suggestions.

Could we perhaps get a simplified dot-point list of all the proposed options so far? Then we can vote on which one we all like the most.

**Comment by quicksketch** [comment]
I'm going to work on rerolling the PR, maintaining the new separate column but with the name `published`. There is some other feedback on the PR itself I'll update as well. Would still love to see this in 1.10, with follow-ups as needed.

**Comment by BWPanda** [comment]
Thanks @quicksketch!

I personally prefer option 2 and believe that if we're going to do this, we should do it well and do it right (i.e. it may be more work, and more changes, but the more properly Backdrop works and the less what-the-heck moments for our users, the better).

IMO.

**Comment by docwilmot** [comment]
> I'm going to work on rerolling the PR

I'm happy with the suggested new direction. @quicksketch are you preparing a PR or should I update?

**Comment by quicksketch** [comment]
Excellent, thanks @docwilmot! I'll push a commit to your PR and then you can continue. I have done the following:

- Changed "submitted_on" to just "submitted".
- Made the default value of the field include am/pm by default.
- Fixed timezone handling in a few places.

I probably haven't fixed all the timezone issues yet. The thing to watch out for is that using strtotime() or date() will use the OS-level timezone to determine how to convert to/from timestamps (which are always UTC). Because the time shown to the user may be the Backdrop timezone _or_ the user's individual timezone, you have to be careful with conversions so you're always converting to/from that timezone rather than the system timezone. Examples will be in the updated commit.

**Comment by quicksketch** [comment]
I pushed up my commit (https://github.com/backdrop/backdrop/pull/2139/commits/bb1d759ff2cd3afb345f81d794964f3486bfd467) to the PR https://github.com/backdrop/backdrop/pull/2139.

Things that still need doing (and we can decide to do them here or in follow-ups):

- As @herbdool added in the PR, this should be a per-content type setting. Things like forum posts probably don't need this addition.
- If this published date is going to be shown on the front-end in the submitted-by line, there needs to be a way to change this date on the node edit form even after the content is published.
- Change the "submitted by" line to use publish date (if available). Otherwise fallback to created date (this will be the case for draft content that is not scheduled).
- We should fix the update hook so that "created" is copied into "published" for all published content.
- We need to update `Backdrop.behaviors.nodeFieldsetSummaries()` to handle the new radio buttons and published field so the vertical tabs display summaries.
- Double check timezone handling is correct.

Things that could (probably should) be later:

- Remove "created" date from the node form.
- Then we'll have "Authoring information" contain only one field: the username. Consider combining all publishing, date, and author information into a single fieldset instead of having them separate.

**Comment by quicksketch** [comment]
Note: Sorry my commit broke tests (and generally submitting new nodes). I can pick this up Monday if needed.

**Comment by quicksketch** [comment]
https://github.com/backdrop/backdrop/pull/2139 is ready for review.

There is one test that I simply cannot get to pass. Ironically related to the timezone handling I said we need to fix. Testbot seems to get confused between the timezone of the test user and that of the user who is running the tests. Not sure how to fix that test, as everything works perfectly in trying to replicate the problem in the UI.

**Comment by laryn** [comment]
I took a look on the site and I think it's looking good. I tried adjusting the default content type publishing settings, created content with a future publishing date, waited & ran cron, and all worked as expected. Created date updated upon publication to the scheduled date.

One potential improvement that could be a follow-up is to unobtrusively include the server timezone (and/or a "Current server time is XX:XX" notification) after the date/time field so the person who is scheduling it knows how it compares to their local time (which may not be the same). Also, it does pre-fill with the current date/time (as of page load) which can also be a bit of a fail if you don't change it, spend time typing on the page, and then get a "can't schedule a past date" warning.

**Comment by quicksketch** [comment]
> (and/or a "Current server time is XX:XX" notification)

I tried adding something like this but it made the form "noisy". My attempt was to use #field_suffix and put the timezone in there, but then it had this large "America/Los_Angeles" string.

> Also, it does pre-fill with the current date/time (as of page load) which can also be a bit of a fail if you don't change it

Though the nature of scheduling means a user-entered value will _always_ be necessary. We could add some more JS to the page though to update the time value to be the time at which the schedule radio is clicked, but the user is going to have to set a time either way. _shrugs_

**Comment by laryn** [comment]
> Though the nature of scheduling means a user-entered value will always be necessary.

What about a default of `now +1day` instead of `now`?

**Comment by quicksketch** [comment]
That seems reasonable... It does seem the next day would be the most likely time to schedule content. And it is slightly better than starting off with an invalid (past) point in time. I pushed an update to the PR.

**Comment by docwilmot** [comment]
Checked again, marked RTBC.

**Comment by quicksketch** [comment]
I've merged https://github.com/backdrop/backdrop/pull/2139 into 1.x for 1.10! Yay, thank you everyone! Huge props to a large team that built up these suggestions and turned it into a fairly simple implementation: @klonos, @BWPanda, @wesruv, @pablo-romero, @laryn, and @jenlampton. Super kudos to @serundeputy and @docwilmot for creating the implementation. All told this ended up being simpler than I had expected (apologies for nearly taking it off the rails), but it's a huge boost in functionality. Great job!

**Comment by quicksketch** [comment]
A follow-up we may need to complete before 1.10 though: https://github.com/backdrop/backdrop-issues/issues/3073
